discoverer/src/main/java/eu/seaclouds/platform/discoverer/api/FetchIfAPI.java

```diff
@@ -0,0 +1,128 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package eu.seaclouds.platform.discoverer.api;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import eu.seaclouds.platform.discoverer.core.Discoverer;
+import eu.seaclouds.platform.discoverer.core.Offering;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.*;
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+
+
+@Path("/fetchif")
+@Produces(MediaType.APPLICATION_JSON)
+public class FetchIfAPI {
+    /* vars */
+    private Discoverer discoverer;
+    private JSONParser JSONParser;
+    static Logger log = LoggerFactory.getLogger(FetchIfAPI.class);
+
+    public FetchIfAPI(Discoverer discoverer) {
+        this.discoverer = discoverer;
+        this.JSONParser = new JSONParser();
+    }
+
+    @POST
+    public FetchIfRepresentation getOfferingsIf(@QueryParam("constraints") String constraints)
+            throws IOException {
+
+        ArrayList<String> validOfferingIds = new ArrayList<>();
+        ArrayList<String> validOfferings = new ArrayList<>();
+        JSONObject constraintsObject;
+
+        try {
+            constraintsObject = (JSONObject) JSONParser.parse(constraints);
+            Collection<String> offeringIds = discoverer.getAllOfferingIds();
+
+            for (String offeringId : offeringIds) {
+                Offering offering = this.discoverer.fetchOffer(offeringId);
+                if (this.satisfyAllConstraints(constraintsObject, offering)) {
+                    validOfferingIds.add(offeringId);
+                    validOfferings.add(offering.toTosca());
+                }
+            }
+        } catch (ParseException e) {
+            log.error("Cannot parse constraints");
+            log.error(e.getMessage());
+        }
+
+        return new FetchIfRepresentation(validOfferingIds, validOfferings);
+    }
+
+    private boolean satisfyAllConstraints(JSONObject constraintsObject, Offering offering) {
+        String toscaString = offering.toTosca();
+        Set keys = constraintsObject.keySet();
+
+        for (Object key : keys) {
+            String constraintName = (String) key;
+            String constraintValue = (String) constraintsObject.get(key);
+
+            if (!satisfyConstraint(constraintName, constraintValue, toscaString))
+                return false;
+        }
+
+        return true;
+    }
+
+    private boolean satisfyConstraint(String constraintName, String constraintValue, String toscaString) {
```

Comment:
Yes, I agree with you @adriannieto. This was a decision the whole consortium took one year ago during weekly integration meetings, which hampers reusability of the code.